In [1]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt 
%matplotlib inline 

In [3]:
dataset = pd.read_csv("C:\\Users\\Tacos\\OneDrive\\Documentos\\Universidad\\Curso - Datos, ciberseguridad e inteligencia artificial\\Actividades\\13 bill_authentication.csv")

In [4]:
dataset.shape 

(1372, 5)

Nuestro conjunto de datos tiene 1372 registros y 5 atributos

In [5]:
dataset.head() 

Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

**Preparación de los datos**

Dividiremos nuestros datos en atributos y etiquetas, y luego dividiremos los datos resultantes en conjuntos de entrenamiento y de prueba.

In [6]:
#Dividir en atributos y etiquetas

X = dataset.drop('Class', axis=1) 
y = dataset['Class'] 

In [8]:
#Crear SPLIT de 80:20
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20) 

**Entrenamiento**

In [9]:
from sklearn.tree import DecisionTreeClassifier 
classifier = DecisionTreeClassifier() 
classifier.fit(X_train, y_train)

DecisionTreeClassifier()

**Predicciones**

In [10]:
y_pred = classifier.predict(X_test)

**Evaluación**

In [11]:
from sklearn.metrics import classification_report, confusion_matrix 
print(confusion_matrix(y_test, y_pred)) 
print(classification_report(y_test, y_pred)) 

[[130   2]
 [  4 139]]
              precision    recall  f1-score   support

           0       0.97      0.98      0.98       132
           1       0.99      0.97      0.98       143

    accuracy                           0.98       275
   macro avg       0.98      0.98      0.98       275
weighted avg       0.98      0.98      0.98       275



De 275 instancias de prueba, nuestro algoritmo solo clasificó erróneamente 2. Esto representa una precisión del 99 %. 

**Árbol de decisión para regresión**

Sobre los datos:  predecir el consumo de gasolina (en millones de galones) en 48 estados de Estados Unidos, con base en el impuesto a la gasolina (en centavos), el ingreso per cápita (dólares), las carreteras pavimentadas (en millas) y la proporción de población con licencia de conducir.

In [12]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
%matplotlib inline

In [14]:
dataset = pd.read_csv("C:\\Users\\Tacos\\OneDrive\\Documentos\\Universidad\\Curso - Datos, ciberseguridad e inteligencia artificial\\Actividades\\13 petrol_consumption.csv")

In [15]:
dataset.head() 

Petrol_tax  Average_income  Paved_Highways  Population_Driver_licence(%)  \
0         9.0            3571            1976                         0.525   
1         9.0            4092            1250                         0.572   
2         9.0            3865            1586                         0.580   
3         7.5            4870            2351                         0.529   
4         8.0            4399             431                         0.544   

   Petrol_Consumption  
0                 541  
1                 524  
2                 561  
3                 414  
4                 410

In [16]:
dataset.describe()

Petrol_tax  Average_income  Paved_Highways  \
count   48.000000       48.000000       48.000000   
mean     7.668333     4241.833333     5565.416667   
std      0.950770      573.623768     3491.507166   
min      5.000000     3063.000000      431.000000   
25%      7.000000     3739.000000     3110.250000   
50%      7.500000     4298.000000     4735.500000   
75%      8.125000     4578.750000     7156.000000   
max     10.000000     5342.000000    17782.000000   

       Population_Driver_licence(%)  Petrol_Consumption  
count                     48.000000           48.000000  
mean                       0.570333          576.770833  
std                        0.055470          111.885816  
min                        0.451000          344.000000  
25%                        0.529750          509.500000  
50%                        0.564500          568.500000  
75%                        0.595250          632.750000  
max                        0.724000          968.000000

In [18]:
#Dividir los datos en etiquetas y atributos

X = dataset.drop('Petrol_Consumption', axis=1) 
y = dataset['Petrol_Consumption'] 

La variable X contiene todas las columnas del conjunto de datos, excepto la columna 
"Petrol_Consumption", que es la etiqueta. La variable y contiene los valores de la columna 
"Petrol_Consumption", lo que significa que la variable X contiene el conjunto de atributos y la variable y contiene las etiquetas correspondientes.

In [19]:
#Crear SPLIT 80:20
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

In [20]:
from sklearn.tree import DecisionTreeRegressor 
regressor = DecisionTreeRegressor() 
regressor.fit(X_train, y_train) 

DecisionTreeRegressor()

In [21]:
#Predicciones
y_pred = regressor.predict(X_test) 

In [22]:
#Ver precisión
df=pd.DataFrame({'Actual':y_test, 'Predicted':y_pred})
df

Actual  Predicted
29     534      547.0
4      410      414.0
26     577      554.0
30     571      554.0
32     577      554.0
37     704      574.0
34     487      648.0
40     587      525.0
7      467      414.0
10     580      464.0

Nota: El método train_test_split divide los datos aleatoriamente.

**Evaluación del algoritmo**

In [23]:
from sklearn import metrics 
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_pred)) 
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_pred)) 
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

Mean Absolute Error: 60.2
Mean Squared Error: 6446.2
Root Mean Squared Error: 80.28823076889913


**Random Forest**

Usaremos el mismo conjunto de datos de zarigüeyas para predecir el sexo de una zarigüeya 
basándonos en características como el tamaño del vientre, el ancho del cráneo, etc.

Paso 1. – Separar las características y la etiqueta

In [26]:
import pandas as pd 

df = pd.read_csv("C:\\Users\\Tacos\\OneDrive\\Documentos\\Universidad\\Curso - Datos, ciberseguridad e inteligencia artificial\\Actividades\\13 possum.csv") 
df.sample(5, random_state=44) 

case  site    Pop sex  age  hdlngth  skullw  totlngth  taill  footlgth  \
88    89     7  other   m  6.0     97.7    58.4      84.5   35.0      64.4   
39    40     2    Vic   f  3.0     91.0    55.0      84.5   36.0      72.8   
92    93     7  other   m  3.0     89.2    54.0      82.0   38.0      63.8   
7      8     1    Vic   f  6.0     94.8    57.6      91.0   37.0      72.7   
98    99     7  other   f  3.0     93.3    56.2      86.5   38.5      64.8   

    earconch   eye  chest  belly  
88      46.2  14.4   29.0   30.5  
39      51.4  13.6   27.0   30.0  
92      44.9  12.8   24.0   31.0  
7       53.9  14.5   29.0   34.0  
98      43.8  14.0   28.0   35.0

In [28]:
#Eliminar datos faltantes (NA)
df = df.dropna() 

In [29]:
# eliminemos las columnas innecesarias y almacenemos los datos de las características y las etiquetas en variables separadas

X = df.drop(["case", "site", "Pop", "sex"], axis=1) 
y = df["sex"] 

**Paso 2. – Entrenamiento de nuestro modelo de bosque aleatorio**

In [31]:
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=44) # Aquí asignamos el 30 % (test_size=0.3) de nuestros datos a las características de entrenamiento (X_train) y las etiquetas (y_train), mientras que el resto se destina a los datos de prueba (X_test, y_test)
from sklearn.ensemble import RandomForestClassifier 
rf_model = RandomForestClassifier(n_estimators=50, random_state=44) 
rf_model.fit(X_train, y_train) 

RandomForestClassifier(n_estimators=50, random_state=44)

**Paso 3. – Hacer predicciones con nuestro modelo**

(m significa masculino, f significa 
femenino)

In [32]:
predictions = rf_model.predict(X_test) 
predictions 

array(['f', 'm', 'm', 'm', 'f', 'f', 'm', 'm', 'm', 'm', 'm', 'f', 'm',
       'm', 'm', 'm', 'm', 'f', 'm', 'm', 'm', 'f', 'f', 'm', 'm', 'm',
       'm', 'm', 'm', 'm', 'f'], dtype=object)

Comprobar las probabilidades asignadas por el modelo a cada predicción con predict_proba()

Cada array contiene dos probabilidades, ya que tenemos dos categorías para predecir: macho o hembra. El valor de la izquierda muestra la probabilidad predicha de pertenecer a la categoría de hembra, y el segundo valor muestra la misma probabilidad para pertenecer a la categoría de macho. 

In [33]:
rf_model.predict_proba(X_test) 

array([[0.5 , 0.5 ],
       [0.26, 0.74],
       [0.36, 0.64],
       [0.3 , 0.7 ],
       [0.66, 0.34],
       [0.54, 0.46],
       [0.48, 0.52],
       [0.42, 0.58],
       [0.18, 0.82],
       [0.34, 0.66],
       [0.48, 0.52],
       [0.54, 0.46],
       [0.4 , 0.6 ],
       [0.28, 0.72],
       [0.48, 0.52],
       [0.36, 0.64],
       [0.4 , 0.6 ],
       [0.54, 0.46],
       [0.44, 0.56],
       [0.06, 0.94],
       [0.32, 0.68],
       [0.56, 0.44],
       [0.66, 0.34],
       [0.46, 0.54],
       [0.18, 0.82],
       [0.46, 0.54],
       [0.34, 0.66],
       [0.26, 0.74],
       [0.2 , 0.8 ],
       [0.1 , 0.9 ],
       [0.5 , 0.5 ]])

In [34]:
rf_model.classes_

array(['f', 'm'], dtype=object)

Revisar la importancia de cada característica para predecir el sexo de una zarigüeya

In [35]:
rf_model.feature_importances_

array([0.05237143, 0.1461762 , 0.08878178, 0.11322144, 0.07150935,
       0.16099193, 0.11990223, 0.11082213, 0.05286153, 0.08336197])

feature_importances_ refleja el orden de las columnas en tu dataframe de características, así que solo tienes que relacionar sus valores con los nombres de las características, así: 

In [38]:
importances = rf_model.feature_importances_ 
columns = X.columns 
i = 0 
while i < len(columns): 
    print(f" The importance of feature '{columns[i]}' is {round(importances[i]*100, 2)}%,") 
    i +=1

 The importance of feature 'age' is 5.24%,
 The importance of feature 'hdlngth' is 14.62%,
 The importance of feature 'skullw' is 8.88%,
 The importance of feature 'totlngth' is 11.32%,
 The importance of feature 'taill' is 7.15%,
 The importance of feature 'footlgth' is 16.1%,
 The importance of feature 'earconch' is 11.99%,
 The importance of feature 'eye' is 11.08%,
 The importance of feature 'chest' is 5.29%,
 The importance of feature 'belly' is 8.34%,


Observe que la longitud de la cabeza (footlgth) tiene el mayor poder de predicción 
(16.1 %) para el sexo de una zarigüeya.

**Introducir nuevos datos para crear una nueva predicción**

In [41]:
new_possum = [[7.0, 83.2, 54.3, 81.0, 37.0, 70.0, 46.3, 14.7, 25.0, 32.0]] 
rf_model.predict(new_possum) 	

C:\Users\Tacos\.conda\envs\ai_security\lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


array(['m'], dtype=object)